ocaml/libs/zstd/zstd.ml

```diff
@@ -20,3 +20,23 @@ module Default = Xapi_compression.Make (struct
 
   let decompress_options = ["--decompress"; "--stdout"]
 end)
+
+module Adaptive = Xapi_compression.Make (struct
+  (** Path to the zstd binary *)
+  let executable = "/usr/bin/zstd"
+
+  (* default compression level is 3 on a scale 1..19 with lower=faster
+     and higher=more compression. By default 1 thread is used *)
+  let compress_options = ["--adapt=min=1,max=4"; "--threads=2"]
```

Comment:
Apparently contrary to the manpage `min=0` is accepted too, might be interesting to see if it has any effect.
If you have a machine and a dump of the stream you'd be compressing you can check the relative speeds this way:
```
zstd -b0 -e4 yourfile
```

And it should print the speed and resulting size for each one.

There is also `lz4` which although doesn't support adaptive compression supports the same `-b` and `-e` flags

Reply:
Although from a quick test level 0 results in weird speed (slower than 1!), so maybe best avoided and use just the values from the manpage as you've done here:
```
zstd -b0 -e2 foo.tar; lz4 -b0 -e2 foo.tar;
 0#foo.tar           :  30853120 ->   6520977 (4.731), 165.5 MB/s , 656.7 MB/s
 1#foo.tar           :  30853120 ->   7315797 (4.217), 243.4 MB/s , 817.8 MB/s
 2#foo.tar           :  30853120 ->   6971746 (4.425), 213.0 MB/s , 721.8 MB/s
Benchmarking levels from 0 to 2
 0#foo.tar           :  30853120 ->  10587545 (2.914), 394.3 MB/s ,2302.8 MB/s
 1#foo.tar           :  30853120 ->  10587545 (2.914), 396.0 MB/s ,2308.5 MB/s
 2#foo.tar           :  30853120 ->  10587545 (2.914), 398.3 MB/s ,2230.9 MB/s```

Reply:
Try moving `--threads=2` to the `--fast` section.